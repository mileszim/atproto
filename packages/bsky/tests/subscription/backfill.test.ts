import AtpAgent from '@atproto/api'
import { SeedClient } from '../seeds/client'
import { runTestEnv, TestEnvInfo } from '@atproto/dev-env'
import { forSnapshot, processAll } from '../_util'
import usersBulk from '../seeds/users-bulk'
import { chunkArray, DAY, wait } from '@atproto/common'
import { ids } from '@atproto/api/src/client/lexicons'

describe('repo subscription backfill', () => {
  let testEnv: TestEnvInfo
  let agent: AtpAgent
  let pdsAgent: AtpAgent
  let sc: SeedClient
  let dids: string[]

  beforeAll(async () => {
    testEnv = await runTestEnv({
      dbPostgresSchema: 'subscription_backfill',
      bsky: { repoSubBackfillConcurrency: 9 },
    })
    agent = new AtpAgent({ service: testEnv.bsky.url })
    pdsAgent = new AtpAgent({ service: testEnv.pds.url })
    sc = new SeedClient(pdsAgent)
    await wait(100) // allow pending sub to be established
    await testEnv.bsky.sub.destroy()
    await usersBulk(sc, 50)

    // For consistent ordering
    dids = Object.keys(sc.dids)
      .sort()
      .map((handle) => sc.dids[handle])

    // Place all events out of the backfill period...
    await testEnv.pds.ctx.db.db
      .updateTable('repo_seq')
      .set({
        sequencedAt: new Date(Date.now() - 2 * DAY).toISOString(),
      })
      .execute()
    // ...except one, which we should pick-up from.
    await updateProfile(pdsAgent, dids[0], { displayName: 'ping' })
  })

  afterAll(async () => {
    await testEnv.close()
  })

  it('ingests all repos via backfill.', async () => {
    // Ensure no profiles have been indexed
    const profilesBefore = await getAllProfiles(agent, dids)
    expect(profilesBefore).toEqual([])

    // Process backfill
    testEnv.bsky.sub.resume()
    await processAll(testEnv, 60000)

    // Check all backfilled profiles
    const profilesAfter = await getAllProfiles(agent, dids)
    expect(dids).toEqual(profilesAfter.map((p) => p.did))
    expect(forSnapshot(profilesAfter)).toMatchSnapshot()
  })

  it('continues processing after backfill.', async () => {
    await updateProfile(pdsAgent, dids[0], { displayName: 'updated' })
    await processAll(testEnv)
    const { data: profile } = await agent.api.app.bsky.actor.getProfile({
      actor: dids[0],
    })
    expect(profile.displayName).toEqual('updated')
  })

  async function getAllProfiles(agent: AtpAgent, dids: string[]) {
    const profileChunks = await Promise.all(
      chunkArray(dids, 25).map(async (chunk) => {
        const { data } = await agent.api.app.bsky.actor.getProfiles({
          actors: chunk,
        })
        return data.profiles
      }),
    )
    return profileChunks.flat()
  }

  async function updateProfile(
    agent: AtpAgent,
    did: string,
    record: Record<string, unknown>,
  ) {
    return await agent.api.com.atproto.repo.putRecord(
      {
        repo: did,
        collection: ids.AppBskyActorProfile,
        rkey: 'self',
        record,
      },
      { headers: sc.getHeaders(did), encoding: 'application/json' },
    )
  }
})
