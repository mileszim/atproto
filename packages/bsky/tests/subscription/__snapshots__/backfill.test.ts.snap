// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`repo subscription backfill ingests all repos via backfill. 1`] = `
Array [
  Object {
    "did": "user(0)",
    "displayName": "ping",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "adrienne49.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(2)/cids(0)@jpeg",
    "did": "user(1)",
    "displayName": "Cody Berge",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "alec-bergnaum.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(4)/cids(0)@jpeg",
    "did": "user(3)",
    "displayName": "Winifred Harber",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "alexander-hickle.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(6)/cids(0)@jpeg",
    "did": "user(5)",
    "displayName": "Carlton Abernathy IV",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "aliya-hodkiewicz.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(7)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "ariel-krajcik.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(8)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "bianka33.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(9)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "brycen-smith.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(10)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "carlos6.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(12)/cids(0)@jpeg",
    "did": "user(11)",
    "displayName": "Rachel Kshlerin",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "cayla-marquardt39.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(13)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "charles-spencer.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(15)/cids(0)@jpeg",
    "did": "user(14)",
    "displayName": "Kelvin Klein",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "chauncey-klein.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(16)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "dewitt-wunsch.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(18)/cids(0)@jpeg",
    "did": "user(17)",
    "displayName": "Miss Danielle Weber",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "dock-mann91.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(20)/cids(0)@jpeg",
    "did": "user(19)",
    "displayName": "Alberta Fay",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "elouise28.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(22)/cids(0)@jpeg",
    "did": "user(21)",
    "displayName": "Dr. Lowell DuBuque",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "elta48.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(23)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "fidel-rath.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(24)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "fredrick-mueller.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(26)/cids(0)@jpeg",
    "did": "user(25)",
    "displayName": "Myron Wolf",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "gerda-marquardt.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(27)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "gina97.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(29)/cids(0)@jpeg",
    "did": "user(28)",
    "displayName": "Kelli Schinner III",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "herman-gleichner95.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(30)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "hilario84.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(31)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "jamil-batz.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(33)/cids(0)@jpeg",
    "did": "user(32)",
    "displayName": "Miss Terrell Ziemann",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "janiya48.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(34)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "javon24.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(36)/cids(0)@jpeg",
    "did": "user(35)",
    "displayName": "Patrick Sawayn",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "jeffrey-sawayn87.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(38)/cids(0)@jpeg",
    "did": "user(37)",
    "displayName": "Eileen Stroman",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "jermain-smith.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(40)/cids(0)@jpeg",
    "did": "user(39)",
    "displayName": "Erica Mertz",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "jonathan-green.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(41)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "joshuah-parker11.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(42)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "kaycee66.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(44)/cids(0)@jpeg",
    "did": "user(43)",
    "displayName": "Miss Floyd McClure",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "kayli-bode.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(45)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "kelton-nitzsche43.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(46)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "kiera97.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(47)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "kim41.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(48)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "leann49.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(49)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "lennie-strosin.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(51)/cids(0)@jpeg",
    "did": "user(50)",
    "displayName": "Karl Bosco IV",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "leonel-koch43.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(53)/cids(0)@jpeg",
    "did": "user(52)",
    "displayName": "Sally Funk",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "magnus53.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(54)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "marietta-zboncak.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(56)/cids(0)@jpeg",
    "did": "user(55)",
    "displayName": "Mrs. Clifford Mertz",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "murphy35.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(57)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "nicolas-krajcik10.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(58)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "ola-oconnell.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(59)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "polly-shanahan45.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(60)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "raleigh-metz.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(62)/cids(0)@jpeg",
    "did": "user(61)",
    "displayName": "Pablo Schmidt",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "reina-runte33.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(63)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "roderick-dibbert.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(65)/cids(0)@jpeg",
    "did": "user(64)",
    "displayName": "Yolanda VonRueden",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "rodger-maggio24.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(67)/cids(0)@jpeg",
    "did": "user(66)",
    "displayName": "Karl Goyette",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "rosamond38.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(69)/cids(0)@jpeg",
    "did": "user(68)",
    "displayName": "Tanya Denesik",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "silas77.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(70)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "sven70.test",
    "labels": Array [],
    "postsCount": 0,
  },
  Object {
    "did": "user(71)",
    "followersCount": 0,
    "followsCount": 0,
    "handle": "tressa-senger.test",
    "labels": Array [],
    "postsCount": 0,
  },
]
`;
